packages/core/test/actor.test.ts

```diff
@@ -1105,4 +1105,79 @@ describe('actors', () => {
       })
       .start();
   });
+
+  // https://github.com/statelyai/xstate/issues/2507
+  it('should not crash on sync child completion before stopping', () => {
+    const childMachine = createMachine({
+      initial: 'idle',
+      states: {
+        idle: {
+          always: [
+            {
+              target: 'stopped'
+            }
+          ]
+        },
+        stopped: {
+          type: 'final'
+        }
+      }
+    });
+
+    const parentMachine = createMachine<{
+      child: ActorRefFrom<typeof childMachine> | null;
+    }>(
+      {
+        context: {
+          child: null
+        },
+        entry: 'setup'
+      },
+      {
+        actions: {
+          setup: assign({
+            child: () => spawn(childMachine)
+          })
+        }
+      }
+    );
+    const service = interpret(parentMachine);
+    expect(() => {
+      service.start();
+    }).not.toThrow();
+  });
+
+  it('should not crash on immediate spawned promise completion', () => {
+    const parentMachine = createMachine<{
+      child: ActorRef<any> | null;
+    }>({
+      context: {
+        child: null
+      },
+      entry: assign({
+        child: () => spawn(new Promise((res) => res(null)))
+      })
+    });
+    const service = interpret(parentMachine);
+    expect(() => {
+      service.start();
+    }).not.toThrow();
+  });
```

Comment:
Since `then` callbacks are always invoked asynchronously this couldn't really throw for this specific scenario. While completion might be immediate and sync - the notification to the parent isn't.

A better test (or an additional one) would be to use a "thenable" here that would actually notify the parent synchronously about its completion:
```suggestion
  it('should not crash on immediate spawned promise completion', () => {
    const parentMachine = createMachine<{
      child: ActorRef<any> | null;
    }>({
      context: {
        child: null
      },
      entry: assign({
        child: () => spawn({ then: (fn) => fn(null) } as any)
      })
    });
    const service = interpret(parentMachine);
    expect(() => {
      service.start();
    }).not.toThrow();
  });
```

Note that this particular test wouldn't fail, even with a thenable. That's only because we cast thenables to promises here:
https://github.com/statelyai/xstate/blob/8d7cbc36432c26fb33e485dbddbe41b784de3fc6/packages/core/src/interpreter.ts#L959

Without this cast this test would fail and this is something that we should have a test for.